### PR TITLE
✨Add unit tests for compute service

### PR DESCRIPTION
**Pull request by mdbooth**
**What this PR does / why we need it**:
This PR adds unit tests for CreateInstance and getImageID in the compute service.

It is split out from #1030, which was getting too large.

It also highlights a bug with port cleanup which will be addressed by a separate PR.

/hold

### CODE REVIEW

File: pkg/cloud/services/networking/service.go
Comment (macaptain) [inline pkg/cloud/services/networking/service.go:62]: tiny typo that was already in the code, not a blocker
```suggestion
		return nil, fmt.Errorf("error retrieving project id: %v", err)
```

File: pkg/cloud/services/compute/service.go
Comment (macaptain) [inline pkg/cloud/services/compute/service.go:33]: I guess it's a philosophical difference at what point a Client becomes a Service. :smile: I think this is fine, the networking service is definitely a wrapper around the networking client. If the bare compute client serves its purpose, it makes sense to have `computeService` and `networkingService` named consistently in this struct.

File: pkg/cloud/services/provider/provider.go
Comment (macaptain) [inline pkg/cloud/services/provider/provider.go:182]: Good call moving this block of code into this function.

File: pkg/cloud/services/compute/instance.go
Comment (macaptain) [inline pkg/cloud/services/compute/instance.go:40]: :partying_face:

File: pkg/cloud/services/compute/instance.go
Comment (macaptain) [inline pkg/cloud/services/compute/instance.go:44]: Rather than using a global var, did you consider keeping this as a `const`, adding `retryInterval` as an argument to the function, and using the value of the const when the function is called? Then this value can be maintained as constant in the codebase but you can choose what to set it as in the tests.

File: pkg/cloud/services/compute/instance_test.go
Comment (macaptain) [inline pkg/cloud/services/compute/instance_test.go:519]: Mostly we have the networkingService covered by tests in the networking package now, I guess this is more to add some kind of integration test for both packages.

**Review by macaptain — APPROVED** [review]
Amazing work on the tests. I can see that they're all running and passing in the CI log. Good to see that they caught a couple of bugs around port deletion as well.

File: pkg/cloud/services/compute/service.go
Comment (mdbooth) [reply pkg/cloud/services/compute/service.go:33]: Yeah, it's a client. This is easy to fix so I'll update.

File: pkg/cloud/services/provider/provider.go
Comment (mdbooth) [reply pkg/cloud/services/provider/provider.go:182]: I think we're going to have to fix this in order to support application credentials, btw, which is another reason to centralise it.

cc @pierreprinetti

File: pkg/cloud/services/compute/instance_test.go
Comment (mdbooth) [reply pkg/cloud/services/compute/instance_test.go:519]: Honestly I don't want to get too philosophical on what's a unit test and what's an integration test. I've been burned before by the maintenance burden and general uselessness of puritanically-scoped unit tests, so I've tried to keep it simple here.

* The scope of the test is the exported method (CreateInstance/DeleteInstance).
* The test boundary is what we can reasonably mock, which is currently the gophercloud calls.

If we were able to mock network service that might also be a practical boundary, but to be honest I quite the emergent benefit of revealing all the external API calls we're making, and they haven't felt *too* onerous yet.

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go]: I've added https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1061/commits/75ba1a99d206156d0f0889edc14a301f1b8027d8 to see if I like this better than the global. Have a look and tell me what you think. I'm personally undecided on which version is nicer, so I'm very much interested in a second opinion!

I'll squash this commit if we decide it's better, so:

/hold

File: pkg/cloud/services/compute/instance_test.go
Comment (macaptain) [reply pkg/cloud/services/compute/instance_test.go:519]: Yes, I do like that the tests show up all the API calls to OpenStack.

File: pkg/cloud/services/compute/instance.go
Comment (macaptain) [reply pkg/cloud/services/compute/instance.go]: I could be a bit paranoid about global variables, but I like passing the retry interval into the function just a little better. It has the drawback of adding a function argument which isn't absolutely essential, but I like that the dependency on the value is made clear in the function signature and that we maintain the original value as a const. What do you think? Maybe one day if more arguments are added to the function, we can tidy up by passing a config struct.

Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go]: Ok, I'll squash the new commit into the existing one.

Thanks for the reviews!

**Comment by mdbooth** [comment]
I think these are good to go.

/hold cancel

**Comment by jichenjc** [comment]
/approve

works for me, thanks for the PR~

**Pull request merged**
